Jsonlogic roundtrip fix

Tupsu-jy opened the pull request:
So I made a long [post](https://github.com/ukrbublik/react-awesome-query-builder/issues/1084) on what I have done in the issue comment thread and I see you have seen it as well so I dont repeat myself too much here.

As for the expansion of this logic to cover all operators. I cant do that for now (i really need to work on my own stuff) but in few weeks i might be able to do that. That is, if that kind of major reworking of how jsonlogic export import works is okay with you? I mean it would be pretty **major reworking** and some of what i have written here is done with the idea of its use being expanded soon.

I reworked some tests because I felt like they were basically testing to make sure that code does not work properly. Some other tests were reworked just because with this change for some operators "!" is part of their definition and that changes how things are interpreted in import, I also left one test commented out because it does not pass currently. That test imo exposes an identical issue in spel import as existed in jsonlogic. I did not look into that really because i really should be working on some other stuff now.

I also have some uncertainty on how things are supposed to work inside "array groups"? Basically the "! preprocessing" while loop inside convertOp. I feel like that area needs some thinking on how it is supposed to work, what its supposed to achieve and when etc. For example should it really be done always? But then i dont really understand why it even exists so i should probably not comment too much on it now. There are some imperfections with my solutions when it comes to this still. For example if it has ! ->and->in then nothing is reversed. I dont know if this really even matters. But i dont want to think about this any further as i really need to get back to working on my own stuff for a while now.

Please let me know if you have some issues with what i have done here.

Tupsu-jy commented:
One line of new code not tested apparently. I hope its okay

ukrbublik commented:
Thanks for your PR 🙏 
Haven't verified it completely yet but idea looks great.
But I noticed that tests for `reverseOperatorsForNot ` will fail after fixing config to actually include `reverseOperatorsForNot: true`
https://github.com/ukrbublik/react-awesome-query-builder/pull/1092/files#diff-6cd85ea43bf1abc61fe406886051f47bbb0c75b57ea2c482df12d181ceea003eR492

Tupsu-jy commented:
Huh true... I guess i was kind of in a haste when i was making these tests. However after quick look I think its just the test that is wrong. What it actually returns is the behavior we want, isnt it?

Tupsu-jy commented:
I fixed the test that was failing and expanded tests somewhat also. Also with the !group negation handling I changed it little bit so that single rule negation groups are negated and removed in import. This is okay i think? Its not a pretty solution but it keeps behavior as it should, for now. I can't really spend much time on this for the next 2 weeks, but after that I could return to this topic and expand the use of template matching logic as much as possible. If that is okay?

Also in my opinion exclamations "!" should not be used to refer to negation groups with one rule at all. This current solution works, but i don't feel like it is ideal. Using "!" for both a single rule negation group and as part of an operators definition is something that can be made to work as has been done here, but its the kind of thing that requires a bunch of special cases, is confusing and is prone to causing bugs in the future. If user wants to create a negation group with one rule wouldnt it be more reasonable to represent it as negation of group (ie. ! -> and)? Removing these groups can condense the json a little bit but that just does not sound like very significant difference.

ukrbublik commented:
I've added 2 small changes:
1. Perform this logic only if `reverseOperatorsForNot == false`
https://github.com/ukrbublik/react-awesome-query-builder/pull/1092/commits/95653464acda8caf33e13e2996815c0172c37ffc#diff-918c51667e2d6a7a67b8118b7495d780e2d9ea31eb91b223c8c74714473de419R163-R170
2. Similar for this logic: https://github.com/ukrbublik/react-awesome-query-builder/pull/1092/commits/14e34fdd45bfcbe9b39be72cf41673d106fbc7b9#diff-1c79cf696987a6ce0d9d91dd34d11cce2c2c02d72960ee06d7e083b6b397142fR837-R841
Enabling it only if `reverseOperatorsForNot == true`

It requires one test change: https://github.com/ukrbublik/react-awesome-query-builder/pull/1092/commits/95653464acda8caf33e13e2996815c0172c37ffc#diff-8edbc6b83d1d4666c36679b38c5bf0b90d1dc369eab63effe107520a43f72e23R921
I think it's how `reverseOperatorsForNot: true` should behave - simplify JsonLogic (without roundtrip).

I think `reverseOperatorsForNot` options and corresponding logic to support it could be removed in future.

If you have further ideas to improve import/export feel free to open a new PR.
